libr/core/cmd_search.c

```diff
@@ -1549,6 +1550,141 @@ static void do_esil_search(RCore *core, struct search_parameters *param, const c
+	arr = malloc (bsize);
+	if (!buf || !arr) {
+		eprintf ("Cannot allocate %d bytes\n", bsize);
+		free(buf);
+		free(arr);
+		return;
+	}
+	r_cons_break_push (NULL, NULL);
+	r_list_foreach (param->boundaries, iter, map) {
+		ut64 from = map->itv.addr;
+		ut64 to = r_itv_end (map->itv);
+		for (i = 0, at = from; at < to; at++, i++) {
+			if (r_cons_is_breaked ()) {
+				break;
+			}
+			if (i >= (bsize - 32)) {
+				i = 0;
+			}
+			if (!i) {
+				r_core_read_at (core, at, buf, bsize);
+			}
+			ret = r_anal_op (core->anal, &aop, at, buf + i, bsize - i);
+			curpos = idx++ % (MAXINSTR + 1);
+			previnstr[curpos] = ret; // This array holds prev n instr size + cur instr size
+			if (ret) {
+				bool match = false;
+				if (aop.type == R_ANAL_OP_TYPE_SWI) {
+					match = true;
+				}
+				if (match) {
+					// This for calculating no of bytes to be subtracted , to get n instr above syscall
+					int nbytes = 0;
+					int nb_opcodes = MAXINSTR;
+					SUMARRAY (previnstr, nb_opcodes, nbytes);
+					curpc = at - (nbytes - previnstr[curpos]);
+					r_reg_set_value (core->dbg->reg, r, curpc);
+					// This loop is for emulating n instr above the syscall
+					for (j = 0; curpc < at; curpc++, j++) {
+						if (j >= (bsize - 32)) {
+							j = 0;
+						}
+						if (!j) {
+							r_core_read_at (core, curpc, arr, bsize);
+						}
+						inslen = r_anal_op (core->anal, &aop, curpc, arr + j, bsize - j);
+						if (inslen) {
+							bool tmatch = false;
+							if (isnonlinear (aop.type)) {
+								tmatch = true;	
+							}
+							int incr = (core->search->align > 0)? core->search->align - 1:  inslen - 1;
+							if (incr < 0) {
+								incr = minopcode;
+							}	
+							j += incr;
+							curpc += incr;
+							if (tmatch) {	// skip the instr
+								r_reg_set_value (core->dbg->reg, r, curpc + 1);
+							} else {	// step instr
+								r_core_esil_step (core, UT64_MAX, NULL, NULL);
+							}
+						}
+					}
+					int off = r_debug_reg_get (core->dbg, a0);
+					RSyscallItem *item = r_syscall_get (core->anal->syscall, off, -1);
+					if (item) {
+						r_cons_printf ("0x%08"PFMT64x" %s\n", at, item->name);	
+					}
+					memset (previnstr, 0, sizeof (previnstr)); // clearing the buffer
+					r_reg_set_value (core->dbg->reg, reg_a0, -2); // clearing register A0
+					if (searchflags) {
+						char *flag = r_str_newf ("%s%d_%d", searchprefix, kwidx, count);
+						r_flag_set (core->flags, flag, at, ret);
+					}
+					if (*param->cmd_hit) {
+						ut64 here = core->offset;
+						r_core_seek (core, at, true);
+						r_core_cmd (core, param->cmd_hit, 0);
+						r_core_seek (core, here, true);
+					}
+					count++;
+					if (search->maxhits > 0 && count >= search->maxhits) {
+						break;
+					}
+				}
+				int inc = (core->search->align > 0)? core->search->align - 1: ret - 1;
+				if (inc < 0) {
+					inc = minopcode;
+				}
+				i += inc;
+				at += inc;
+			}
+		}
+	}
+	r_anal_esil_free (esil);
+	r_anal_esil_stack_free (core->anal->esil);
```

Comment:
ouch, nope. dont do that. this will break the current esil stack. not the one you have setup for that specific analysis

Reply:
also, for that emulation we shouldnt be doing memwrites or stack changes at all, so it shouldnt be necessary to free the stack